### PR TITLE
Inital commit.

### PR DESCRIPTION
**Pull request by AalexxJjamess**

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -56,14 +56,18 @@ Purpose:  This project will show you the difference between member functions and
...
-    T(<#type name#> v, const char* <#variable name#>)   //1
-    //2
-    //3
+    T(float v, const char* charPointer)
+    {
+        value = v;
+        chrPtr = charPointer;
+    }
+    float value;
+    std::string chrPtr;
```
Comment (matkatmusic) [inline main.cpp:65]: Good, but use the member initializer list instead.

File: main.cpp [hunk 2]
```diff
@@ -73,29 +77,29 @@ struct <#structName1#>                                //4
 
 struct U
 {
-    float <#name1#> { 0 }, <#name2#> { 0 };
-    <#returnType#> <#memberFunction#>(<#type name#>* <#updatedValue#>)      //12
+    float a { 0 }, b { 0 };
+    float uMemberFunc(float* updatedValue)      //12
     {
-        
+        return a * b;
```
Comment (matkatmusic) [inline main.cpp:83]: You need to copy the guts of your `ValuePrinterStruct::valuePrinterFunc` to this function and replace all instances of `that->` accordingly.  The purpose of this is to show you the difference between static functions and member functions.
- [ ] revise.

File: main.cpp [hunk 3]
```diff
@@ -73,29 +77,29 @@ struct <#structName1#>                                //4
...
 {
-    static <#returntype#> <#staticFunctionA#>(U* that, <#type name#>* <#updatedValue#> )        //10
+    static float valuePrinterFunc(U* that, float* updatedValue)        //10
     {
-        std::cout << "U's <#name1#> value: " << that-><#name1#> << std::endl;
-        that-><#name1#> = <#updatedValue#>;
-        std::cout << "U's <#name1#> updated value: " << that-><#name1#> << std::endl;
-        while( std::abs(that-><#name2#> - that-><#name1#>) > 0.001f )
+        std::cout << "U's a value: " << that->a << std::endl;
+        that->a = *updatedValue;
```
Comment (matkatmusic) [inline main.cpp:92]: What do we always do before we use a pointer?
- [ ] revise.

File: main.cpp [hunk 4]
```diff
@@ -56,14 +56,18 @@ Purpose:  This project will show you the difference between member functions and
...
 };
 
-struct <#structName1#>                                //4
+struct Comparator                                //4
 {
-    <#type name#> compare(<#type name#> a, <#type name#> b) //5
+    T* compare(T* a, T* b) //5
     {
         if( a->value < b->value ) return a;
         if( a->value > b->value ) return b;
```
Comment (matkatmusic) [inline main.cpp:73]: What do we always do before we use a pointer?
- [ ] revise.

File: main.cpp [hunk 5]
```diff
@@ -115,19 +119,29 @@ struct <#structname2#>
...
+    auto* smaller = f.compare(&firstT, &secondT);
+        
+    if(smaller == nullptr)
+        {
+            std::cout << "They are equal." << std::endl;
+        }
+    else
+        {
+            std::cout << "the smaller one is << " << smaller->chrPtr << std::endl;
+        }
```
Comment (matkatmusic) [inline main.cpp:135]: Fix the indentation.
Also, check your logic again.  What happens if you pass a `nullptr` to compare?  Are they equal if you do that?
- [ ] revise.

File: main.cpp (hunk 1 above)
Comment (AalexxJjamess) [reply main.cpp:65]: Done.

File: main.cpp (hunk 5 above)
Comment (AalexxJjamess) [reply main.cpp:135]: Done.

File: main.cpp (hunk 4 above)
Comment (AalexxJjamess) [reply main.cpp:73]: Check that it isn't nullptr. Added.

File: main.cpp [hunk 6]
```diff
@@ -69,37 +65,54 @@ struct Comparator                                //4
...
     T* compare(T* a, T* b) //5
     {
-        if( a->value < b->value ) return a;
-        if( a->value > b->value ) return b;
+        if (a && b != nullptr)
+        {
+            if( a->value < b->value ) return a;
+            if( a->value > b->value ) return b;
+        }
         return nullptr;
```
Comment (matkatmusic) [inline main.cpp]: Fix this nullptr check logic. 
This is interpreted by the compiler as:
`if( a )` **and** `if( b != nullptr )`
you are not checking if `a` is not a `nullptr`.  you are only checking `b`
You need to look up order of operations.   When in doubt, be explicit about what you're comparing. 
- [ ] revise.

File: main.cpp [hunk 7]
```diff
@@ -69,37 +65,54 @@ struct Comparator                                //4
...
 
 struct ValuePrinterStruct
 {
     static float valuePrinterFunc(U* that, float* updatedValue)        //10
     {
-        std::cout << "U's a value: " << that->a << std::endl;
-        that->a = *updatedValue;
-        std::cout << "U's a updated value: " << that->a << std::endl;
-        while( std::abs(that->b - that->a) > 0.001f )
+        if (that && updatedValue != nullptr)
```
Comment (matkatmusic) [inline main.cpp:103]: See previous comments.
- [ ] revise

File: main.cpp (hunk 6 above)
Comment (AalexxJjamess) [reply main.cpp]: Done.

**Pull request merged**